prometheus_client/core.py

```diff
@@ -235,7 +240,60 @@ def get(self):
       with self._lock:
           return self._value
 
-_ValueClass = _MutexValue
+
+def _MultiProcessValue(__pid=os.getpid()):
+    pid = __pid
+    samples = {}
+    samples_lock = Lock()
+
+    class _ShelveValue(object):
+        '''A float protected by a mutex backed by a per-process shelve.'''
+
+        _multiprocess = True
+
+        def __init__(self, typ, metric_name, name, labelnames, labelvalues, multiprocess_mode='', **kwargs):
+            with samples_lock:
+                if typ == 'gauge':
+                    file_prefix = typ + '_' +  multiprocess_mode
+                else:
+                    file_prefix = typ
+                if file_prefix not in samples:
+                    filename = os.path.join(os.environ['prometheus_multiproc_dir'], '{0}_{1}.db'.format(file_prefix, pid))
+                    samples[file_prefix] = shelve.open(filename)
+            self._samples = samples[file_prefix]
+            self._key = json.dumps((metric_name, name, labelnames, labelvalues))
+            self._value = self._samples.get(self._key, 0.0)
+            self._samples[self._key] = self._value
+            self._samples.sync()
+            self._lock = Lock()
+
+        def inc(self, amount):
+            with self._lock:
+                self._value += amount
+                self._samples[self._key] = self._value
+                self._samples.sync()
+
+        def set(self, value):
+            with self._lock:
+                self._value = value
+                self._samples[self._key] = self._value
+                self._samples.sync()
+
+        def get(self):
+            with self._lock:
+                return self._value
+
+    return _ShelveValue
+
+
+# Should we enable multi-process mode?
+# This needs to be chosen before the first metric is constructed,
+# and as that may be in some arbitrary library the user/admin has
+# no control over we use an enviroment variable.
+if 'prometheus_multiproc_dir' in os.environ:
```

Comment:
This seems like a reasonable idea.

> Somehow configuring the system explicitly at process startup seems like the best path

That won't work, as by the time your code is running all this code has likely already been imported and many metrics already exist. That's why I'm using environment variables, as they're set at `exec()` time before any python code has a chance to run.